src/Catalog/Helpers/Db2CatalogProjection.cs

```diff
@@ -19,29 +23,70 @@ public Db2CatalogProjection(PackageContentUriBuilder packageContentUriBuilder)
 
             return new FeedPackageDetails(
                 packageContentUri,
-                dataRecord.ReadDateTime(Db2CatalogProjectionColumnNames.Created).ForceUtc(),
-                dataRecord.ReadNullableUtcDateTime(Db2CatalogProjectionColumnNames.LastEdited),
-                listed ? dataRecord.ReadDateTime(Db2CatalogProjectionColumnNames.Published).ForceUtc() : Constants.UnpublishedDate,
+                dataReader.ReadDateTime(Db2CatalogProjectionColumnNames.Created).ForceUtc(),
+                dataReader.ReadNullableUtcDateTime(Db2CatalogProjectionColumnNames.LastEdited),
+                listed ? dataReader.ReadDateTime(Db2CatalogProjectionColumnNames.Published).ForceUtc() : Constants.UnpublishedDate,
                 packageId,
                 normalizedPackageVersion,
-                hideLicenseReport ? null : dataRecord[Db2CatalogProjectionColumnNames.LicenseNames]?.ToString(),
-                hideLicenseReport ? null : dataRecord[Db2CatalogProjectionColumnNames.LicenseReportUrl]?.ToString());
+                hideLicenseReport ? null : dataReader[Db2CatalogProjectionColumnNames.LicenseNames]?.ToString(),
+                hideLicenseReport ? null : dataReader[Db2CatalogProjectionColumnNames.LicenseReportUrl]?.ToString(),
+                deprecationInfo);
+        }
+
+        public PackageDeprecationItem ReadDeprecationInfoFromDataReader(DbDataReader dataReader)
+        {
+            if (dataReader == null)
+            {
+                throw new ArgumentNullException(nameof(dataReader));
+            }
+
+            var deprecationReasons = new List<string>();
+            var deprecationStatusValue = dataReader.ReadInt32OrNull(Db2CatalogProjectionColumnNames.DeprecationStatus);
+
+            if (!deprecationStatusValue.HasValue)
+            {
+                return null;
+            }
+
+            var deprecationStatus = (PackageDeprecationStatus)deprecationStatusValue.Value;
+
+            foreach (var deprecationStatusFlag in Enum.GetValues(typeof(PackageDeprecationStatus)).Cast<PackageDeprecationStatus>())
+            {
+                if (deprecationStatusFlag == PackageDeprecationStatus.NotDeprecated)
+                {
+                    continue;
+                }
+
+                if (deprecationStatus.HasFlag(deprecationStatusFlag))
+                {
+                    deprecationReasons.Add(deprecationStatusFlag.ToString());
+                }
+            }
+
+            var alternatePackageVersion = dataReader.ReadStringOrNull(Db2CatalogProjectionColumnNames.AlternatePackageVersion);
+
+            return new PackageDeprecationItem(
+                deprecationReasons,
+                dataReader.ReadStringOrNull(Db2CatalogProjectionColumnNames.DeprecationMessage),
+                dataReader.ReadStringOrNull(Db2CatalogProjectionColumnNames.AlternatePackageId),
+                alternatePackageVersion == null ? "*" : $"[{alternatePackageVersion}, {alternatePackageVersion}]");
```

Comment:
If `AlternatePackageId` and `AlternatePackageVersion` are both null, it looks like you'll be passing `null` and `*` into this method, [which will cause `PackageDeprecationItem`'s constructor to throw](https://github.com/NuGet/NuGet.Services.Metadata/blob/master/src/Catalog/PackageDeprecationItem.cs#L41).

I'd also suggest making sure the unit tests cover:
null ID and null version
not null ID and null version
not null ID and not null version

Reply:
Good catch! This was hidden by a faulty test setup.
Fixed in https://github.com/NuGet/NuGet.Services.Metadata/pull/542/commits/592665ecec20b491925cc20e76805c167608fc3b